### ISSUE TITLE
Any way to know how much ram is free after compilation / execution?

### ISSUE THREAD

**Issue by xlar54** [cc65/cc65#710]
Ive got a program that is starting to crash on the 64 and I believe Im starting to hit memory limits.  First, should I be using dynamic memory more or less, and secondly, is there any way to determine if Im simply at a ram limit?  Perhaps a freemem() type of function?  Maybe there is a wiki on this kind of info?

**Comment by oliverschmidt** [comment]
> First, should I be using dynamic memory more or less [...] ?

I'm having a hard time to understand that question. So I'm just guessing...

In general there's no benefit in dynamic memory. If you have e.g. a program that does at some point a malloc(1000) and later a malloc(500) you're always better of with two static arrays. The latter is better both from the performance and better from the memory consumption - because if you program doesn't use dynamic memory at all then the dynamic memory management code (malloc/free/...) doesn't become part of your program at all.

Dynamic memory is only useful if you want use memory at different time during the run time of your program for different things. Or if the memory usage is balanced between different things you don't know the size of at compile time. Think of searching a text for a list of words: Maybe you have a large list and a rather short text or the opposite.

> is there any way to determine if Im simply at a ram limit? Perhaps a freemem() type of function? Maybe there is a wiki on this kind of info?

From that perspective there are basically three types of memory relevant:

1. Everything the linker know it's size of. If that doesn't fit into your RAM the linker will bail out. If you want to know how close you already are to that point then let the linker create a map file and check how much memory is left between the end of the BSS segment ($664E in the example below) ... 

```
Segment list:
-------------
Name                   Start     End    Size  Align
----------------------------------------------------
[...]
BSS                   0047CF  00664E  001E80  00001
[...]
```

... and the end address of the memory reserved of the BSS for your target. For the C64 that is c64.cfg saying:

```
SYMBOLS {
[...]
    __STACKSIZE__: type = weak, value = $0800; # 2k stack
    __HIMEM__:     type = weak, value = $D000;
}
MEMORY {
[...]
    BSS: file = "", start = __ONCE_RUN__, size = __HIMEM__ - __STACKSIZE__ - __ONCE_RUN__;
}
```

So the end of the BSS memory is at $C800 meaning that from the linker perspective there are in the example above $61B2 bytes free.

2. It's unknown how much stack you program will need during run time. However, instead of just pretending that it doesn't need any stack at all you rather tell the linker how much stack it should presume you might use. For the C64 that's by default 2kB. If you use less then you "waste" memory. If you use more then you - after using the free memory - overwrite global/static variables of your program. Note that there's http://cc65.github.io/doc/cc65.html#pragma-check-stack (and a compiler command line option for the same thing). However, I personally have never used it so I can't tell you anything more than what's written there.

3. Dynamic memory (aka the heap). It's location depends on your target. For the C64 http://cc65.github.io/doc/c64.html#s3 says that it's located between the BSS and the stack - that's the very area I classified as "free" above. That means that the stack size given to the linker (those 2kB by default) doesn't only determine when the linker will complain about a memory overflow. It also directly influences your heap size. In the example above the initial heap size is exactly $61B2 bytes. If the heap is exhausted then a malloc() returns NULL.

To summarize:

> starting to crash on the 64 and I believe Im starting to hit memory limits

The only scenario where this can happen is if you use more stack than you told the linker you'll use. You're supposed to be able to check that using the `--check-stack` compiler option. Or you try to increase the stack size. If you still have no linker errors and no malloc's returning with NULL then you still have "memory left" meaning that it's rather unlikely that you have a "stack overwrite issue".

> Perhaps a freemem() type of function?

If you want to know after some malloc's and free's how much heap you currently have, call http://cc65.github.io/doc/funcref.html#_heapmemavail

**Comment by xlar54** [comment]
thanks oliver - that all helps and is very thorough.

**Comment by oliverschmidt** [comment]
Thanks for the positive feedback :-)
